Fix change detection for databags

skywalk7 opened the pull request:
Summary:
For changes detection we don't care if file was added or modified or have any other status provided by VCS, we only need to understand if file was modified or deleted. So using same logic as in BetweenMeals::Changes::Role

This would fix an issue with Grocery Delivery uploads when adding new databag.

jaymzh commented:
> This would fix an issue with Grocery Delivery uploads when adding new databag.

Huh? We support uploading databags. I'm confused as to the issue here.

skywalk7 commented:
> > This would fix an issue with Grocery Delivery uploads when adding new databag.
> 
> Huh? We support uploading databags. I'm confused as to the issue here.

I think issue was introduced with HG support. 
`BetweenMeals::Repo::Svn` returns only `:deleted` or `:modified`:
```
:status => Regexp.last_match(1) == 'D' ? :deleted : :modified,
```
`BetweenMeals::Repo::Git` returns `:modified` if file was added or modified:
```
when 'A'
  # A path
  {
    :status => :modified,
    :path => parts[1],
  }
...
when /^M(?:\d*)/
  # M<numbers> path
  {
    :status => :modified,
    :path => parts[1],
  }
```
`BetweenMeals::Repo::Hg` returns `:modified` or `:added`:
```
when 'A'
  {
    :status => :added,
    :path => parts[1],
  }
...
when 'M'
  {
    :status => :modified,
    :path => parts[1],
  }
```
Changing `BetweenMeals::Repo::Hg` is another solution, but I think it is cleaner to update change detection, also it would implement same logic as already used for roles.

Why this problem was not detected earlier? Databags are not used now, but when they were used, Chef was running on Git or SVN.

jaymzh commented:
Well  `git` is already doing this conversion for us. So it seems like we expect the plugins to do it. So if we do this, then we should let `git` pass through `:added`. Or we should modify `hg` to not pass through `added`. Either is fine.

We should also add some comments to `initialize` to be clear what we expect plugins to pass in and what we do with it.

skywalk7 commented:
> So it seems like we expect the plugins to do it.

No, we don't expect it, look at the code for roles.

>  So if we do this, then we should let `git` pass through `:added`. 

This could break things that I and you are not aware of.

> Or we should modify `hg` to not pass through `added`. Either is fine.

Also could cause some side effects, but if you insist I can do it.

> We should also add some comments to `initialize` to be clear what we expect plugins to pass in and what we do with it.

Makes sense, let me update the code.

jaymzh commented:
> No, we don't expect it, look at the code for roles.

Well, roles does this same thing:

```ruby
      def initialize(file, role_dir)
        @status = file[:status] == :deleted ? :deleted : :modified
        @name = self.class.name_from_path(file[:path], role_dir)
      end
```

So we don't expect it for role changes but do for others. We should be consistent across the code-base. Either the object-type classes normalize the `:status` **everywhere** (roughly what you're doing), or we make all VCS classes normalize their `:status` output. The lack of consistency is what got us in this situtation, so if we're gonna fix it, lets define the expectations of the API better and make everything consistent.

We have a ton of tests for a reason, so we can fix bugs like this and have a good idea that we're not breaking existing behavior "I don't wanna fix the code because it may break stuff" is never a good reason not to fix things.

skywalk7 commented:
I've updated code and tests. Unfortunately commits came in separately.

skywalk7 commented:
@jaymzh Can you please take a look?

jaymzh commented:
Sorry, was out of town. Kicked off tests. Will look at code a bit later today.

jaymzh commented on lib/between_meals/changes/change.rb:
```diff
@@ -24,13 +24,22 @@ module BetweenMeals
   module Changes
     # Common functionality
     class Change
+      ALLOWED_STATUSES = [:modified, :deleted].freeze
```
```suggestion
      # Since we either need to upload or delete, we only accept two statuses.
      # VCSs will differentiate between various kinds of modifies, adds, etc.
      # so instead of handling all possibilities here, we expect the caller to
      # collapse them into `:modified` or `:deleted`.
      ALLOWED_STATUSES = [:modified, :deleted].freeze
```

jaymzh reviewed (changes requested):
This looks great. Just add a comment here for our future selves. ;)

jaymzh reviewed (approved):
Beautiful! Thanks! It doesn't just fix a bug, it makes the API cleaner, and the code base and tests more resilient. Thanks!
